pennylane/vqe/vqe.py

```diff
@@ -129,6 +130,70 @@ def __str__(self):
 
         return "\n+ ".join(terms)
 
+    def __matmul__(self, H):
+
+        coeffs1 = self.coeffs.copy()
+        terms1 = self.ops.copy()
+
+        coeffs2 = H.coeffs
+        terms2 = H.ops
+
+        coeffs = [i[0] * i[1] for i in itertools.product(coeffs1, coeffs2)]
+        term_list = itertools.product(terms1, terms2)
+
+        terms = [qml.operation.Tensor(i[0], i[1]) for i in term_list]
+
+        return qml.Hamiltonian(coeffs, terms)
```

Comment:
What happens if we compute the tensor product of Hamiltonians acting on the same wires? We should probably throw an error, right?

Reply:
Yes, possibly, however I didn't add this because currently the qml.Hamiltonian object allows us to define Hamiltonians with tensor products of terms acting on the same wire (even though errors are thrown if we try to use it in another method later on)

Reply:
The long term goal is for `@` on the same wire to act as matrix multiplication. `@` will be somewhat contextual, and its action as a tensor product or matrix product dependent on the arguments.

E.g.,

```pycon
>>> qml.PauliX(0) @ qml.PauliX(1)
Tensor(PauliX(wires=[0]), PauliX(wires=[1]))
>>> qml.PauliX(0) @ qml.PauliX(0)
qml.Identity(0)
```

Reply:
Just for my own knowledge: why can't we represent non-Hermitian operators? I haven't seen anywhere that it is enforced that the coefficients are real, but I may be missing something. In any case it's not a big deal, I'm mostly curious about what would happen if we computed `qml.PauliX(0) @ qml.PauliX(0)`

Reply:
@ixfoduap The Hamiltonian class checks that coefficients are real upon initialization.

Reply:
However, we can define non-Hermitian operators as long as the coefficients are real (for example, if we pass `qml.PauliZ(0) @ qml.PauliY(0)`, no error will be thrown)

Reply:
> @ixfoduap The Hamiltonian class checks that coefficients are real upon initialization.

Yep, this is the reason. This is probably historical; the `Hamiltonian` class was created for VQE, and we didn't want the user passing a non-Hermitian Hamiltonian as an expectation value.